ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,276 @@
+    findPossiblyLeakedStateDirs();
+    findPossiblyLeakedTranientTopics();
+  }
+
+  public void addCleanupTask(final Callable<Boolean> task) {
+    cleanupTasks.add(task);
+  }
+
+  private void findPossiblyLeakedStateDirs() {
+    final String stateDir = this.stateDir;
+    final File folder = new File(stateDir);
+    final File[] listOfFiles = folder.listFiles();
+
+    if (listOfFiles == null) {
+      return;
+    }
+
+    for (File f: listOfFiles) {
+      final String fileName = f.getName();
+      final Matcher filenameMatcher = TRANSIENT_PATTERN.matcher(fileName);
+
+      if (filenameMatcher.find()) {
+        final String leakingQueryId = filenameMatcher.group(0);
+        if (allTransientQueriesEverCreatedIds.contains(leakingQueryId)
+                && isTerminatedQuery(leakingQueryId)) {
+          LOG.info("{} is leaking state directories. Adding it to the cleanup queue.",
+                  leakingQueryId);
+
+          addCleanupTask(new TransientQueryStateCleanupTask(
+                  leakingQueryId,
+                  stateDir)
+          );
+        }
+      }
+    }
+  }
+
+  private void findPossiblyLeakedTranientTopics() {
+    for (String topic : this.serviceContext
+            .getTopicClient()
+            .listTopicNames()) {
+      final Matcher topicNameMatcher = TRANSIENT_PATTERN.matcher(topic);
+
+      if (topicNameMatcher.find()) {
+        final String leakingQueryId = topicNameMatcher.group(0);
+        if (allTransientQueriesEverCreatedIds.contains(leakingQueryId)
+                && isTerminatedQuery(leakingQueryId)) {
+          LOG.info("{} is leaking topics. Adding it to the cleanup queue.", leakingQueryId);
+          addCleanupTask(new TransientQueryTopicCleanupTask(
+                  serviceContext,
+                  leakingQueryId
+          ));
+        }
+      }
+    }
+  }
+
+  public void recordTransientQueryOnStart(final TransientQueryMetadata queryMetadata) {
+    this.allTransientQueriesEverCreated.add(queryMetadata);
+    this.allTransientQueriesEverCreatedIds.add(queryMetadata.getQueryApplicationId());
+  }
+
+  public static class TransientQueryTopicCleanupTask implements Callable<Boolean> {
+
+    private final ServiceContext serviceContext;
+    private final String appId;
+
+    public TransientQueryTopicCleanupTask(
+            final ServiceContext serviceContext,
+            final String appId
+    ) {
+      this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
+      this.appId = Objects.requireNonNull(appId, "appId");
+    }
+
+    @Override
+    public Boolean call() {
+      try {
+        LOG.info("Deleting topics for transient query: {}", appId);
+        serviceContext.getTopicClient().deleteInternalTopics(appId);
+      } catch (final Exception e) {
+        LOG.warn("Failed to cleanup topics for transient query: {}", appId, e);
+        return false;
+      }
+
+      final Set<String> orphanedTopics = serviceContext
```

Comment:
Maybe it's not a big deal in practice, but each time a query is completed, it will do this call to Kafka to list all of the topics.  Also, deleteInternalTopics lists topics as well.  If you have a bunch of queued up Tasks, they each will do the identical call, at all once.  In practice we don't do many push queries, but you can imagine that you could queue up 300 of these in a ten minute period.

It might make sense to use the batch delete and only call listTopics once.

Anyway, we currently use this approach for `QueryCleanupService`, so it's probably fine, but maybe something to keep in mind.